compiler/src/dotty/tools/dotc/transform/ShortcutImplicits.scala

```
@@ -61,82 +73,42 @@ class ShortcutImplicits extends MiniPhase with IdentityDenotTransformer { thisPh
-  }
-
-  /** A new `m$direct` method to accompany the given method `m` */
-  private def newDirectMethod(sym: Symbol)(implicit ctx: Context): Symbol = {
-    val direct = sym.copy(
-      name = DirectMethodName(sym.name.asTermName).asInstanceOf[sym.ThisName],
-      flags = sym.flags | Synthetic,
-      info = directInfo(sym.info))
-    if (direct.allOverriddenSymbols.isEmpty) direct.resetFlag(Override)
-    direct
-  }
 
   /** The direct method `m$direct` that accompanies the given method `m`.
     *  Create one if it does not exist already.
     */
   private def directMethod(sym: Symbol)(implicit ctx: Context): Symbol =
-    if (sym.owner.isClass) {
-      val direct = sym.owner.info.member(DirectMethodName(sym.name.asTermName))
-        .suchThat(_.info matches directInfo(sym.info)).symbol
-      if (direct.maybeOwner == sym.owner) direct
-      else newDirectMethod(sym).enteredAfter(thisPhase)
-    }
-    else directMeth.getOrElseUpdate(sym, newDirectMethod(sym))
+    if (sym.owner.isClass) shortcutMethod(sym, thisPhase)
+    else directMeth.getOrElseUpdate(sym, newShortcutMethod(sym))
 
   /** Transform `qual.apply` occurrences according to rewrite rule (2) above */
   override def transformSelect(tree: Select)(implicit ctx: Context) =
     if (tree.name == nme.apply &&
         defn.isImplicitFunctionType(tree.qualifier.tpe.widen) &&
-        shouldBeSpecialized(tree.qualifier.symbol)) {
+        needsImplicitShortcut(tree.qualifier.symbol)) {
       def directQual(tree: Tree): Tree = tree match {
         case Apply(fn, args)     => cpy.Apply(tree)(directQual(fn), args)
         case TypeApply(fn, args) => cpy.TypeApply(tree)(directQual(fn), args)
         case Block(stats, expr)  => cpy.Block(tree)(stats, directQual(expr))
         case tree: RefTree =>
+          def rewire(tp: Type, sym: Symbol) = tp match {
+            case tp: NamedType => tp.prefix.select(sym)
+            case _ => sym.termRef
```

Could you explain when the last case is reachable?

---

I'll try to eliminate the last case. It's better to fail fast if we misunderstand something.